src/charm.py

```diff
@@ -12,92 +12,292 @@
+                try:
+                    resource.remove()
+                except ApiException as exc:
+                    if exc.status == 404:
+                        pass  # Resource already deleted.
+                    else:
+                        raise exc
         """
-        current = self.config["thing"]
-        if current not in self._stored.things:
-            logger.debug("found a new thing: %r", current)
-            self._stored.things.append(current)
+        config.load_kube_config()
+
+        core_api = client.CoreV1Api()
+        auth_api = client.RbacAuthorizationV1Api()
+        storage_api = client.StorageV1Api()
+        app_api = client.AppsV1Api()
 
-    def _on_fortune_action(self, event):
-        """Just an example to show how to receive actions.
+        return [
+            Secret(core_api, "csi-rbd-secret", self.K8S_NS),
+            ServiceAccount(core_api, "rbd-csi-nodeplugin", self.K8S_NS),
+            ServiceAccount(core_api, "rbd-csi-provisioner", self.K8S_NS),
+            ClusterRole(auth_api, "rbd-csi-nodeplugin"),
+            ClusterRole(auth_api, "rbd-csi-nodeplugin-rules"),
+            ClusterRole(auth_api, "rbd-external-provisioner-runner-rules"),
+            ClusterRole(auth_api, "rbd-external-provisioner-runner"),
+            ClusterRoleBinding(auth_api, "rbd-csi-nodeplugin"),
+            ClusterRoleBinding(auth_api, "rbd-csi-provisioner-role"),
+            Role(auth_api, "rbd-external-provisioner-cfg", self.K8S_NS),
+            RoleBinding(auth_api, "rbd-csi-provisioner-role-cfg", self.K8S_NS),
+            StorageClass(storage_api, "ceph-ext4"),
+            StorageClass(storage_api, "ceph-xfs"),
+            Service(core_api, "csi-metrics-rbdplugin", self.K8S_NS),
+            Service(core_api, "csi-rbdplugin-provisioner", self.K8S_NS),
+            Deployment(app_api, "csi-rbdplugin-provisioner", self.K8S_NS),
+            ConfigMap(core_api, "ceph-csi-config", self.K8S_NS),
+            ConfigMap(core_api, "ceph-csi-encryption-kms-config", self.K8S_NS),
+            DaemonSet(app_api, "csi-rbdplugin", self.K8S_NS),
+        ]
 
-        TEMPLATE-TODO: change this example to suit your needs.
-        If you don't need to handle actions, you can remove this method,
-        the hook created in __init__.py for it, the corresponding test,
-        and the actions.py file.
+    def update_stored_state(self, config_name: str, stored_state_name: str) -> bool:
+        """Update value in stored stated based on the value from config.
 
-        Learn more about actions at https://juju.is/docs/sdk/actions
+        Since ConfigChangedEvent does not provide information about which config options
+        were updated, this convenience method allows you to pass name of the config
+        option and name of the corresponding variable in StoredState. If the values
+        do not match, StoredState variable is updated with new value and `True` is
+        returned.
```

Comment:
This sounds either like conscious decision or that there's some other limitation. Maybe because the `ConfigChangedEvent` does not necessarily have to be triggered by direct call to `juju config`([soo config-changed section](https://juju.is/docs/sdk/events)). But maybe I can open a conversation with the Ops team about this.